week2/homework/exercises/1-blog-API/server.js

```diff
@@ -1,10 +1,117 @@
-const express = require('express')
+const express = require('express');
 const app = express();
- 
+const fs = require('fs');
+const path = require('path');
+const { title } = require('process');
 
 // YOUR CODE GOES IN HERE
-app.get('/', function (req, res) {
-  res.send('Hello World')
-})
- 
-app.listen(3000)
\ No newline at end of file
+app.use(express.json());
+
+function isInvalid(req) {
+  if (
+    typeof req.body.title == 'undefined' ||
+    typeof req.body.content == 'undefined'
+  ) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+// creating a blog
+app.post('/blogs', (req, res) => {
+  // How to get the title and content from the request??
+  creatBlog(req, res);
+});
+
+// updating a blog:
+app.put('/posts/:title', (req, res) => {
+  updateBlog(req, res);
+});
+// deleting a blog:
+app.delete('/blogs/:title', (req, res) => {
+  deleteBlog(req, res);
+});
+// reading a blog:
+
+app.get('/blogs/:title', (req, res) => {
+  readBlog(req, res);
+});
+
+// read all posts"
+app.get('/blogs', (req, res) => {
+  //   // how to get the file names of all files in a folder??
+  res.body = [];
+  fs.readdir('./posts', (err, files) => {
+    files.forEach((file) => {
+      res.body.push({ title: file });
+    });
+    res.send(res.body);
+  });
+});
+
+function creatBlog(req, res) {
+  if (isInvalid(req)) {
+    res.status(400);
+    res.send('invalid request');
+    return;
+  }
+  let title = req.body.title;
+  let content = req.body.content;
+  fs.writeFileSync(path.join(__dirname, './posts', title), content);
+  res.end('ok');
+}
+
+function updateBlog(req, res) {
+  // How to get the title and content from the request?
+  let title = req.params.title;
+  let content = req.body.content;
+  // What if the request does not have a title and/or content?
+  if (isInvalid(req)) {
+    res.status(400);
+    res.send('invalid request');
+    return;
+  }
+
+  if (fs.existsSync(path.join(__dirname, './posts', title))) {
+    fs.writeFileSync(path.join(__dirname, './posts', title), content);
+    res.end('ok');
+  } else {
+    // Send response with error message
+    res.status(404);
+    res.send('This post does not exist!');
+    return;
+  }
```

Comment:
I am a big fan of early returns and hate nesting my code so this is how I'd write it
```suggestion
  if ( ! fs.existsSync(path.join(__dirname, './posts', title))) {
    // Send response with error message
    res.status(404);
    res.send('This post does not exist!');
    return;
 }
 fs.writeFileSync(path.join(__dirname, './posts', title), content);
 res.end('ok');
 ```